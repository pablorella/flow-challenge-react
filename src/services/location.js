import request from '../utils/request';

export default {
  fetch: () => {
    const url = 'https://ip-api.com/json/';
    return request(url)
  }
}