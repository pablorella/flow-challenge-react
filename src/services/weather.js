import request from '../utils/request';

export default {
  fetchSelectedCity: city => {
    const url = `https://api.openweathermap.org/data/2.5/forecast/daily?q=${city}&units=metric&&appid=${process.env.REACT_APP_ACCESS_KEY}`;
    return request(url)
  }
}