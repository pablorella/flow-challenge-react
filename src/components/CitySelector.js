import React, {useState, useEffect} from 'react';
import styled from 'styled-components';
import location from '../services/location';
import weather from '../services/weather';
import Loading from './Loading';
import WeatherDayCard from './WeatherDayCard';
import Button from './Button';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  select{
    width: 20%;
    font-size: 1rem;
  }
`


const CardsContainer = styled.div`
  display: flex;
  width: 100%;
`
const CitySelector = () => {

  const getCityWeather = (list) => {

    const firstDay =  {
      date: new Date(list[0].dt * 1000),
      icon: list[0].weather[0].icon,
      description: list[0].weather[0].main,
      min:  Math.floor(list[0].temp.min),
      max: Math.floor(list[0].temp.max),
      humidity: list[0].humidity
    }
    const secondDay =  {
      date: new Date(list[1].dt * 1000 ),
      icon: list[1].weather[0].icon,
      description: list[1].weather[0].main,
      min:  Math.floor(list[1].temp.min),
      max: Math.floor(list[1].temp.max),
      humidity: list[1].humidity
    }
    const thirdDay =  {
      date: new Date(list[2].dt * 1000),
      icon: list[2].weather[0].icon,
      description: list[2].weather[0].main,
      min:  Math.floor(list[2].temp.min),
      max: Math.floor(list[2].temp.max),
      humidity: list[2].humidity
    }
    const fourthDay =  {
      date: new Date(list[3].dt * 1000),
      icon: list[3].weather[0].icon,
      description: list[3].weather[0].main,
      min:  Math.floor(list[3].temp.min),
      max: Math.floor(list[3].temp.max),
      humidity: list[3].humidity
    }
    const fivethDay =  {
      date: new Date(list[4].dt * 1000),
      icon: list[4].weather[0].icon,
      description: list[4].weather[0].main,
      min:  Math.floor(list[4].temp.min),
      max: Math.floor(list[4].temp.max),
      humidity: list[4].humidity
    }

    return [firstDay, secondDay, thirdDay, fourthDay, fivethDay ]
  }

  const [ currentCity, setCurrentCity ] = useState(null)
  const [ selectedCityWeather, setSelectCityWeather ] = useState(null)
  const [ selectedCity, setSelectedCity ] = useState('')
  const [ isFetching, setIsFetching ] = useState(false);

  useEffect( () => {
    const fetchLocation = async() => {
      try {
        const { city } = await location.fetch();
        const { list } = await weather.fetchSelectedCity(city)
        const cityWeather = getCityWeather(list)

        setCurrentCity(city);
        setSelectCityWeather(cityWeather)
        setSelectedCity(city)

      }catch(e){
        alert('there was an error');
      }
    }
     fetchLocation();
  },[])

  const handleClickCheckWeather = async () => {
    setIsFetching(true)
    const { list } = await weather.fetchSelectedCity(selectedCity)
    const cityWeather = getCityWeather(list)
    setIsFetching(false)
    setSelectCityWeather(cityWeather)
  }
  const renderWeatherCards = () => {
    return selectedCityWeather.map( day => {
      return <WeatherDayCard isFetching={isFetching} weather={day} />
    })
  }

  if( !currentCity || !selectedCityWeather ) return <Loading />
  return (
    <Container>
      <h4>Please, select a city</h4>
      <select onChange={(e) => setSelectedCity(e.target.value)}>
        <option value={currentCity}>{currentCity}</option>
        <option value="London">London</option>
        <option value="Paris">Paris</option>
        <option value="Milan">Milan</option>
        <option value="Istambul">Istambul</option>
        <option value="Bangkok">Bangkok</option>
      </select>
      <Button disabled={isFetching} handleOnClick={handleClickCheckWeather} title="Check weather" />
      <CardsContainer>
        { renderWeatherCards() }
      </CardsContainer>
    </Container>
  )
}

export default CitySelector;