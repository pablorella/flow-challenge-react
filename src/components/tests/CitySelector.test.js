import React from 'react';
import { render, cleanup, fireEvent, wait } from '@testing-library/react';
import CitySelector from '../CitySelector';
import { act } from 'react-dom/test-utils';
import location from '../../services/location';
import weather from '../../services/weather'


describe.skip('CitySelector', () => {

  afterEach(() => {
    cleanup();
    jest.clearAllMocks()
  });
  beforeEach( () => {
    location.fetch = jest.fn( () => Promise.resolve( { city: 'La Plata'}));
    weather.fetchSelectedCity = jest.fn( () => Promise.resolve( {
       list: [
         { dt: new Date(),
          weather: [
             { icon: 'md10', main: 'Rain'  }
            ],
            temp : {
              min: '2',
              max: '10'
            },
            humidity: '10'
          },
          { dt: new Date(),
            weather: [
               { icon: 'md10', main: 'Rain'  }
              ],
              temp : {
                min: '2',
                max: '10'
              },
              humidity: '10'
          },
          { dt: new Date(),
            weather: [
              { icon: 'md10', main: 'Rain'  }
            ],
            temp : {
              min: '2',
              max: '10'
            },
            humidity: '10'
          },
          { dt: new Date(),
            weather: [
                { icon: 'md10', main: 'Rain'  }
              ],
              temp : {
                min: '2',
                max: '10'
              },
              humidity: '10'
            },
            { dt: new Date(),
              weather: [
                  { icon: 'md10', main: 'Rain'  }
                ],
                temp : {
                  min: '2',
                  max: '10'
                },
                humidity: '10'
              },
        ]
       }));
    window.alert = jest.fn();
  } )


  it('should matches snapshot', () => {
    const { asFragment } = render(<CitySelector />)
    expect(asFragment()).toMatchSnapshot();
  })

  it('should render the correct content when is loading', async () => {
    location.fetch = jest.fn( () => Promise.resolve( {}));

   let getByTestId
    await act( async() => {
      ( { getByTestId } = await render(<CitySelector />))
    });
    expect(getByTestId('loading')).toBeTruthy();
  })

  it('should render the correct content after fetch location successfully', async () => {

   let getByText
   await act( async() => {
      ( { getByText } = await render(<CitySelector />))
    });
     expect(getByText("La Plata")).toBeInTheDocument();
  })

  it('should render the correct content after fetch location and fails', async () => {
    location.fetch = jest.fn( () => Promise.reject());

    await act( async() => {
       await render(<CitySelector />);
     });
      expect(window.alert).toHaveBeenCalled();
   })

  it('should call the correct action when a city is selected', async () => {

    let getByTestId
    let getByText
    await act( async() => {
       ( { getByTestId, getByText } = await render(<CitySelector />))
     });

     const select = getByTestId('city-select')
     const submitButton = getByText('Check weather');

     await act( async() => {
        await fireEvent.change(select, { target: { value: "London" }});
        await fireEvent.click(submitButton)
    });

     expect(weather.fetchSelectedCity).toHaveBeenCalledWith('London');

   })

})