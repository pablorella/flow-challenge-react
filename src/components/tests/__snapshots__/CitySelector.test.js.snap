// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitySelector should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm ghYKab"
    data-testid="loading"
  >
    <p>
      Loading...
    </p>
  </div>
</DocumentFragment>
`;
